doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,446 @@
+
+**Pros:**
+
+- With a sufficiently high Y, no queries or writes would be blocked.
+- May be helpful if/when we wanted to make the coordinator distributed.
+
+**Cons:**
+
+- Complicated to implement and likely has a large blast radius.
+- Requires [#8426](https://github.com/MaterializeInc/materialize/issues/8426)
+  to be completed first.
+- If X is low enough we will overflow our timestamps at some point in the future. This can be mitigated with a high
+  enough X.
+
+##### Write Ahead Log (WAL)
+
+This approach involves the coordinator buffering all writes to user tables in a durable write ahead log. All writes to
+user tables that occur in the same millisecond can be assigned the same timestamp, as long as there are no reads
+in-between them. Then at the end of the millisecond, or after X milliseconds, all writes are sent to STORAGE in a batch.
+If Materialize crashes/restarts before sending the writes to STORAGE, then it can just read the unsent writes from the
+durable logs after starting back up. This prevents us from having to increment the global timestamp on every write to a
+user table. However, this does not solve the problem if there are reads in-between the writes or if we read from a view
+that's ahead of the wall clock.
+
+**Pros:**
+
+- No queries are blocked from consecutive writes to user tables.
+- A WAL would be useful for many other reasons not in scope for this document. It’s likely that no matter what solution
+  we pick, we’d still want to implement a WAL eventually.
+
+**Cons:**
+
+- Only solves part of the problem.
+- Requires the WAL backend to have a distributed consensus operation.
+
+##### Fast Forward
+
+Timestamps in Materialize are meant to closely track the wall clock, but are not guaranteed to be exactly equal to the
+wall clock. This approach would involve forcibly fast-forwarding the `upper`s of all objects. Whenever an operation
+would push the global timestamp above the `upper` of other objects by some threshold, all `upper`s are jumped to the
+global timestamp’s new value. The threshold could be set to 0 to prevent blocking any queries.
+
+**Pros:**
+
+- No queries are blocked by more than the assigned threshold.
+
+**Cons:**
+
+- Certain operations will require a broadcast to all nodes.
+- I’m not actually sure if this is possible, but if it is possible then it’s probably complicated and has a large blast
+  radius.
+
+## Alternatives
+
+- TODO
+
+## Open Questions
+
+- This document does not go into great detail on fault tolerance and how crashes/restarts affect consistency. All read
+  and write operations are atomic and idempotent so crashing/restarting will not leave the system in an intermediate or
+  invalid state due to read/writes. However, DDLs are not idempotent or atomic. What happens if the system crashes in
+  the middle of DDL? Some of this is being thought about by the command reconciliation work.
+- How does the global timestamp get re-established after a restart?
```

Comment:
There are maybe other solutions here that avoid a durable write on every read like: on startup, ensure the `since` of everything is at least that thing's `upper`, which would guarantee that we can't do reads at a previous time after startup (assuming `upper`s are durably progressed, which I think is a good assumption).

Reply:
It's a good point that there are other possible solutions! I'm not sure I grok this one specifically, but we want to not accidentally respond to things that on restart we might respond to differently. One similar thing to what you said is "initialize the global timestamp to the maximum `upper`" which has other scary properties, but ensures that we haven't responded to anything at further timestamp.